src/wasm_fetcher/registry.rs

```diff
@@ -0,0 +1,56 @@
+use anyhow::{anyhow, Result};
+use async_std::fs::File;
+use async_std::prelude::*;
+use async_trait::async_trait;
+use oci_distribution::Reference;
+use oci_distribution::client::{Client, ClientConfig, ClientProtocol};
+use oci_distribution::secrets::RegistryAuth;
+use std::str::FromStr;
+use tokio_compat_02::FutureExt;
+use url::Url;
+
+use crate::wasm_fetcher::fetcher::Fetcher;
+
+// Struct used to reference a WASM module that is hosted on an OCI registry
+pub(crate) struct Registry {
+  // full path to the WASM module
+  destination: String,
+  // url of the remote WASM module
+  wasm_url: String,
+}
+
+impl Registry {
+    pub(crate) fn new(url: Url) -> Result<Registry> {
+        match url.path().rsplit('/').next() {
+            Some(image_ref) => {
+                let wasm_url = url.to_string();
+                Ok(Registry{
+                    destination: image_ref.into(),
+                    wasm_url: wasm_url
+                        .strip_prefix("registry://")
+                        .map_or(Default::default(), |url| url.into()),
+                })
+            },
+            None => Err(anyhow!(
+                "Cannot infer name of the remote file by looking at {}",
+                url.path()
+            )),
+        }
+    }
+}
+
+#[async_trait]
+impl Fetcher for Registry {
+    async fn fetch(&self) -> Result<String> {
+        let mut client = Client::new(ClientConfig{
+            protocol: ClientProtocol::Https,
+        });
+        let reference = Reference::from_str(self.wasm_url.as_str()).unwrap();
+        let image_content = client.pull_image(&reference, &RegistryAuth::Anonymous).compat().await?.content;
```

Comment:
Can you add a comment explaining why we are using the `compact()` function? Just to leave a note about the discrepancies between tokio versions and what we did to sort them out.